### PR TITLE
Bootstrap patchelf like GnuPG

### PR DESCRIPTION
**Pull request by alalazo**
Remove a custom bootstrapping procedure to use functions in `spack.bootstrap` instead

### CODE REVIEW

**Comment by haampie** [comment]
Would be great if we can finally bootstrap this from relocatable (in the true sense, not in the spack-sense) pre-built binaries.

Should we just build and distribute patchelf with statically linked musl libc? Licences are compatible

**Comment by alalazo** [comment]
Going in the direction you say is in the roadmap for v0.18.0, but this PR is obviously only a tiny step in that direction. To have a `patchelf` that can be just "copied into place" we need to do improve the buildcache mechanism first.

**Comment by haampie** [comment]
Can we fix the error message propagation from the nested bootstrapping business?

```
tmp.23f3os8vPu $ spack spec zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
==> Bootstrapping clingo from pre-built binaries
==> Error: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %gcc target=x86_64" due to the following failures:
    Please run `spack -d spec zlib` for more verbose error messages
```

I suppose it's easier to retrieve this error when we don't do nested bootstrapping in the first place...

If all bootstrapping goes through one code path, can't we bootstrap patchelf from sources as a prerequisite there so that it's easier to report errors and so that we don't need refcounting?

**Comment by haampie** [comment]
Notice it's important to show a proper error message here, cause if it fails in CI you won't run with -d.

**Comment by alalazo** [comment]
> Can we fix the error message propagation from the nested bootstrapping business?

Sure, see 717923b That commit removes the use of `SpackCommand` but leaves for later the overhaul of the code dealing with buildcaches.

> I suppose it's easier to retrieve this error when we don't do nested bootstrapping in the first place...

Avoid nesting is difficult with the current lazy approach of bootstrapping something only when we need it.

The objective of this PR is to bootstrap `patchelf` using code in `spack.bootstrap` instead of custom code. I'd like to arrive at a point where we use:
- `ensure_module_importable_or_raise` for Python modules
- `ensure_executables_in_path_or_raise` for commands executed in a child process

That said, I'm not opposed to change the lazy bootstrapping logic in another PR if that seems sensible to other maintainers too.

File: lib/spack/spack/relocate.py [hunk 1]
```diff
@@ -76,32 +77,16 @@ def __init__(self, old_path, new_path):
...
-
-    # Skip darwin
+    """Return the full path to the patchelf binary, if available, else None."""
     if is_macos:
         return None
 
-    # Install the spec and return its path
-    spec.package.do_install()
-    return exe_path if os.path.exists(exe_path) else None
+    patchelf = executable.which('patchelf')
```
Comment (haampie) [inline lib/spack/spack/relocate.py:83]: Can we do a follow-up PR where executable.which is replaced with getting a spec out of it through external find logic? Cause we should require a lower-bound `@0.13:` on the version number, given that ubuntu:18.04 ships a very buggy patchelf `0.9`.

File: lib/spack/spack/bootstrap.py [hunk 2]
```diff
@@ -758,10 +783,23 @@ def gnupg_root_spec():
...
 def ensure_gpg_in_path_or_raise():
     """Ensure gpg or gpg2 are in the PATH or raise."""
-    ensure_executables_in_path_or_raise(
+    return ensure_executables_in_path_or_raise(
         executables=['gpg2', 'gpg'], abstract_spec=gnupg_root_spec()
     )
 
+
+def patchelf_root_spec():
+    """Return the root spec used to bootstrap GnuPG"""
```
Comment (haampie) [inline lib/spack/spack/bootstrap.py]:
```suggestion
    """Return the root spec used to bootstrap patchelf"""
```

File: lib/spack/spack/bootstrap.py [hunk 3]
```diff
@@ -758,10 +783,23 @@ def gnupg_root_spec():
...
+    return _root_spec('patchelf')
```
Comment (haampie) [inline lib/spack/spack/bootstrap.py]:
```suggestion
    return _root_spec('patchelf@0.13:')
```

File: lib/spack/spack/bootstrap.py [hunk 4]
```diff
@@ -758,10 +783,23 @@ def gnupg_root_spec():
...
+
+
+def ensure_patchelf_in_path_or_raise():
+    """Ensure gpg or gpg2 are in the PATH or raise."""
```
Comment (haampie) [inline lib/spack/spack/bootstrap.py]:
```suggestion
    """Ensure patchelf is in the PATH or raise."""
```
Or just delete the docstring cause it's not adding any new information given the function name.

**Review by haampie — CHANGES_REQUESTED** [review]

File: lib/spack/spack/bootstrap.py (hunk 4 above)
Comment (alalazo) [reply lib/spack/spack/bootstrap.py]: Would you prefer to fix this or delete all of them?

Comment (haampie) [reply lib/spack/spack/bootstrap.py]: For me it's the same, if it's less work to hit commit suggestion feel free ;p

**Comment by alalazo** [comment]
@haampie Can you check this PR again?

**Review by haampie — APPROVED** [review]

**Pull request merged**